File: android/src/main/java/expo/modules/mediaedit/ProjectModel.kt
package expo.modules.mediaedit

/**
 * Kotlin mirror of the TypeScript Project schema in src/project.ts.
 * Both the compiler and the preview view parse the raw JS map once
 * at the bridge boundary and then operate on these structs.
 */

data class ProjectTimeRange(val startMs: Long, val endMs: Long) {
  val durationMs: Long get() = endMs - startMs
  val startUs: Long get() = startMs * 1000L
  val endUs: Long get() = endMs * 1000L
}

sealed class ProjectClipTransition {
  object Cut : ProjectClipTransition()
  data class Fade(val durationMs: Long) : ProjectClipTransition()
  data class FadeToBlack(val durationMs: Long) : ProjectClipTransition()
}

data class ProjectVideoClip(
  val id: String,
  val sourceUri: String,
  val sourceRange: ProjectTimeRange,
  val timelineRange: ProjectTimeRange,
  val transition: ProjectClipTransition,
  val originalVolume: Float,
  val isImage: Boolean,
  val imageDurationMs: Long?,
)

data class ProjectAudioClip(
  val id: String,
  val sourceUri: String,
  val sourceRange: ProjectTimeRange,
  val timelineRange: ProjectTimeRange,
  val volume: Float,
  val trimToVideo: Boolean,
)

sealed class ProjectOverlayClip {
  abstract val id: String
  abstract val timelineRange: ProjectTimeRange?

  data class Text(
    override val id: String,
    val content: String,
    val x: Float,
    val y: Float,
    val anchor: String,
    val textAlign: String,
    val paddingX: Float,
    val paddingY: Float,
    val fontSize: Float,
    val color: String,
    val fontWeight: String,
    val fontStyle: String,
    val fontFamily: String,
    val backgroundColor: String?,
    val cornerRadius: Float,
    val strokeColor: String?,
    val strokeWidth: Float,
    val shadowColor: String?,
    val shadowRadius: Float,
    val shadowOpacity: Float,
    val highlightColor: String?,
    val highlightStart: Int?,
    val highlightLength: Int?,
    val rotation: Float,
    override val timelineRange: ProjectTimeRange?,
  ) : ProjectOverlayClip()

  data class Image(
    override val id: String,
    val uri: String,
    val x: Float,
    val y: Float,
    val width: Float,
    val height: Float,
    val opacity: Float,
    val rotation: Float,
    override val timelineRange: ProjectTimeRange?,
  ) : ProjectOverlayClip()
}

sealed class ProjectTrack {
  abstract val id: String
  data class Video(override val id: String, val clips: List<ProjectVideoClip>) : ProjectTrack()
  data class Audio(override val id: String, val clips: List<ProjectAudioClip>) : ProjectTrack()
  data class Overlay(override val id: String, val items: List<ProjectOverlayClip>) : ProjectTrack()
}

data class Project(
  val id: String,
  val schemaVersion: Int,
  val canvasWidth: Int,
  val canvasHeight: Int,
  val fps: Int,
  val durationMs: Long,
  val tracks: List<ProjectTrack>,
) {
  fun videoClips(): List<ProjectVideoClip> =
    tracks.flatMap { if (it is ProjectTrack.Video) it.clips else emptyList() }
  fun audioClips(): List<ProjectAudioClip> =
    tracks.flatMap { if (it is ProjectTrack.Audio) it.clips else emptyList() }
  fun overlayClips(): List<ProjectOverlayClip> =
    tracks.flatMap { if (it is ProjectTrack.Overlay) it.items else emptyList() }
}

class ProjectParseException(msg: String) : Exception(msg)

object ProjectParser {

  @Suppress("UNCHECKED_CAST")
  fun parse(map: Map<String, Any?>): Project {
    val id = map["id"] as? String ?: throw ProjectParseException("missing id")
    val schemaVersion = (map["schemaVersion"] as? Number)?.toInt() ?: 1
    if (schemaVersion != 1) throw ProjectParseException("unsupported schemaVersion $schemaVersion")

    val canvas = map["canvasSize"] as? Map<String, Any?>
      ?: throw ProjectParseException("missing canvasSize")
    val canvasW = (canvas["width"] as? Number)?.toInt() ?: throw ProjectParseException("invalid canvasSize.width")
    val canvasH = (canvas["height"] as? Number)?.toInt() ?: throw ProjectParseException("invalid canvasSize.height")

    val fps = (map["fps"] as? Number)?.toInt() ?: 30
    val durationMs = (map["durationMs"] as? Number)?.toLong() ?: 0L

    val tracks = (map["tracks"] as? List<Map<String, Any?>>)?.map { parseTrack(it) } ?: emptyList()

    return Project(
      id = id, schemaVersion = schemaVersion,
      canvasWidth = canvasW, canvasHeight = canvasH,
      fps = fps, durationMs = durationMs, tracks = tracks
    )
  }

  @Suppress("UNCHECKED_CAST")
  private fun parseTrack(m: Map<String, Any?>): ProjectTrack {
    val kind = m["kind"] as? String ?: throw ProjectParseException("track missing kind")
    val id = m["id"] as? String ?: throw ProjectParseException("track missing id")
    return when (kind) {
      "video" -> ProjectTrack.Video(id, (m["clips"] as? List<Map<String, Any?>>)?.map { parseVideoClip(it) } ?: emptyList())
      "audio" -> ProjectTrack.Audio(id, (m["clips"] as? List<Map<String, Any?>>)?.map { parseAudioClip(it) } ?: emptyList())
      "overlay" -> ProjectTrack.Overlay(id, (m["items"] as? List<Map<String, Any?>>)?.map { parseOverlay(it) } ?: emptyList())
      else -> throw ProjectParseException("unknown track kind $kind")
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun parseVideoClip(m: Map<String, Any?>): ProjectVideoClip {
    val id = m["id"] as? String ?: "v-${System.nanoTime()}"
    val kind = m["kind"] as? String
    val isImage = kind == "image"
    val sourceUri = m["sourceUri"] as? String ?: ""
    val imageDurationMs = (m["durationMs"] as? Number)?.toLong()
    val srcRange = parseTimeRange(m["sourceRange"]) ?: ProjectTimeRange(0, imageDurationMs ?: 0)
    val tlRange = parseTimeRange(m["timelineRange"])
      ?: throw ProjectParseException("video clip $id missing timelineRange")
    val transition = parseTransition(m["transition"] as? Map<String, Any?>)
    val originalVolume = (m["originalVolume"] as? Number)?.toFloat() ?: 1f
    return ProjectVideoClip(
      id, sourceUri, srcRange, tlRange, transition, originalVolume,
      isImage = isImage, imageDurationMs = imageDurationMs
    )
  }

  @Suppress("UNCHECKED_CAST")
  private fun parseAudioClip(m: Map<String, Any?>): ProjectAudioClip {
    val id = m["id"] as? String ?: "a-${System.nanoTime()}"
    val sourceUri = m["sourceUri"] as? String ?: throw ProjectParseException("audio clip $id missing sourceUri")
    val srcRange = parseTimeRange(m["sourceRange"])
      ?: throw ProjectParseException("audio clip $id missing sourceRange")
    val tlRange = parseTimeRange(m["timelineRange"])
      ?: throw ProjectParseException("audio clip $id missing timelineRange")
    return ProjectAudioClip(
      id, sourceUri, srcRange, tlRange,
      volume = (m["volume"] as? Number)?.toFloat() ?: 1f,
      trimToVideo = m["trimToVideo"] as? Boolean ?: false
    )
  }

  private fun parseOverlay(m: Map<String, Any?>): ProjectOverlayClip {
    val kind = m["kind"] as? String ?: throw ProjectParseException("overlay missing kind")
    val id = m["id"] as? String ?: "o-${System.nanoTime()}"
    val tl = parseTimeRange(m["timelineRange"])
    return when (kind) {
      "text" -> ProjectOverlayClip.Text(
        id = id,
        content = m["content"] as? String ?: "",
        x = (m["x"] as? Number)?.toFloat() ?: 0f,
        y = (m["y"] as? Number)?.toFloat() ?: 0f,
        anchor = m["anchor"] as? String ?: "center",
        textAlign = m["textAlign"] as? String ?: "center",
        paddingX = (m["paddingX"] as? Number)?.toFloat() ?: 12f,
        paddingY = (m["paddingY"] as? Number)?.toFloat() ?: 8f,
        fontSize = (m["fontSize"] as? Number)?.toFloat() ?: 32f,
        color = m["color"] as? String ?: "#FFFFFF",
        fontWeight = m["fontWeight"] as? String ?: "normal",
        fontStyle = m["fontStyle"] as? String ?: "normal",
        fontFamily = m["fontFamily"] as? String ?: "system",
        backgroundColor = m["backgroundColor"] as? String,
        cornerRadius = (m["cornerRadius"] as? Number)?.toFloat() ?: 0f,
        strokeColor = m["strokeColor"] as? String,
        strokeWidth = (m["strokeWidth"] as? Number)?.toFloat() ?: 0f,
        shadowColor = m["shadowColor"] as? String,
        shadowRadius = (m["shadowRadius"] as? Number)?.toFloat() ?: 0f,
        shadowOpacity = (m["shadowOpacity"] as? Number)?.toFloat() ?: 1f,
        highlightColor = m["highlightColor"] as? String,
        highlightStart = (m["highlightStart"] as? Number)?.toInt(),
        highlightLength = (m["highlightLength"] as? Number)?.toInt(),
        rotation = (m["rotation"] as? Number)?.toFloat() ?: 0f,
        timelineRange = tl,
      )
      "image" -> ProjectOverlayClip.Image(
        id = id,
        uri = m["uri"] as? String ?: "",
        x = (m["x"] as? Number)?.toFloat() ?: 0f,
        y = (m["y"] as? Number)?.toFloat() ?: 0f,
        width = (m["width"] as? Number)?.toFloat() ?: 0.2f,
        height = (m["height"] as? Number)?.toFloat() ?: 0.2f,
        opacity = (m["opacity"] as? Number)?.toFloat() ?: 1f,
        rotation = (m["rotation"] as? Number)?.toFloat() ?: 0f,
        timelineRange = tl,
      )
      else -> throw ProjectParseException("unknown overlay kind $kind")
    }
  }

  @Suppress("UNCHECKED_CAST")
  private fun parseTimeRange(raw: Any?): ProjectTimeRange? {
    val m = raw as? Map<String, Any?> ?: return null
    val s = (m["startMs"] as? Number)?.toLong() ?: return null
    val e = (m["endMs"] as? Number)?.toLong() ?: return null
    return ProjectTimeRange(s, e)
  }

  @Suppress("UNCHECKED_CAST")
  private fun parseTransition(m: Map<String, Any?>?): ProjectClipTransition {
    if (m == null) return ProjectClipTransition.Cut
    return when (m["type"] as? String) {
      "fade" -> ProjectClipTransition.Fade((m["durationMs"] as? Number)?.toLong() ?: 500L)
      "fadeToBlack" -> ProjectClipTransition.FadeToBlack((m["durationMs"] as? Number)?.toLong() ?: 500L)
      else -> ProjectClipTransition.Cut
    }
  }
}

File: ios/ProjectModel.swift
import Foundation
import AVFoundation

/// Swift-side mirror of TypeScript Project (src/project.ts). All
/// fields are pure data — nothing here owns AV objects. Both the
/// compiler and the preview view parse a project dictionary once at
/// the bridge boundary and then operate on these structs.

public struct ProjectTimeRange {
  public let startMs: Double
  public let endMs: Double
  public var startTime: CMTime { CMTime(value: CMTimeValue(startMs), timescale: 1000) }
  public var endTime: CMTime { CMTime(value: CMTimeValue(endMs), timescale: 1000) }
  public var duration: CMTime { endTime - startTime }
  public var durationMs: Double { endMs - startMs }
}

public enum ProjectClipTransition {
  case cut
  case fade(durationMs: Double)
  case fadeToBlack(durationMs: Double)
}

public struct ProjectVideoClip {
  public let id: String
  public let sourceUri: String
  public let sourceRange: ProjectTimeRange
  public let timelineRange: ProjectTimeRange
  public let transition: ProjectClipTransition
  public let originalVolume: Double
  /// `kind == "image"` when the source is a still image that needs to
  /// be expanded into a time-bounded AVAsset. Default false.
  public let isImage: Bool
  public let imageDurationMs: Double?
}

public struct ProjectAudioClip {
  public let id: String
  public let sourceUri: String
  public let sourceRange: ProjectTimeRange
  public let timelineRange: ProjectTimeRange
  public let volume: Double
  public let trimToVideo: Bool
}

public enum ProjectOverlayClip {
  case text(ProjectTextOverlayClip)
  case image(ProjectImageOverlayClip)
}

public struct ProjectTextOverlayClip {
  public let id: String
  public let content: String
  public let x: Double
  public let y: Double
  public let anchor: String       // "topLeft" | "center"
  public let textAlign: String    // "left" | "center" | "right"
  public let paddingX: Double
  public let paddingY: Double
  public let fontSize: Double
  public let color: String
  public let fontWeight: String
  public let fontStyle: String
  public let fontFamily: String
  public let backgroundColor: String?
  public let cornerRadius: Double
  public let strokeColor: String?
  public let strokeWidth: Double
  public let shadowColor: String?
  public let shadowRadius: Double
  public let shadowOpacity: Double
  public let highlightColor: String?
  public let highlightStart: Int?
  public let highlightLength: Int?
  public let rotation: Double
  public let timelineRange: ProjectTimeRange?
}

public struct ProjectImageOverlayClip {
  public let id: String
  public let uri: String
  public let x: Double
  public let y: Double
  public let width: Double
  public let height: Double
  public let opacity: Double
  public let rotation: Double
  public let timelineRange: ProjectTimeRange?
}

public enum ProjectTrack {
  case video(id: String, clips: [ProjectVideoClip])
  case audio(id: String, clips: [ProjectAudioClip])
  case overlay(id: String, items: [ProjectOverlayClip])
}

public struct Project {
  public let id: String
  public let schemaVersion: Int
  public let canvasSize: CGSize
  public let fps: Int
  public let durationMs: Double
  public let tracks: [ProjectTrack]
}

// MARK: - Parsing from the JS bridge dict

public enum ProjectParseError: Error {
  case invalid(String)
}

public struct ProjectParser {

  public static func parse(_ dict: [String: Any]) throws -> Project {
    guard let id = dict["id"] as? String else { throw ProjectParseError.invalid("missing id") }
    let schemaVersion = (dict["schemaVersion"] as? Int) ?? (dict["schemaVersion"] as? Double).map { Int($0) } ?? 1
    guard schemaVersion == 1 else { throw ProjectParseError.invalid("unsupported schemaVersion \(schemaVersion)") }
    guard let canvasDict = dict["canvasSize"] as? [String: Any],
          let w = (canvasDict["width"] as? Double) ?? (canvasDict["width"] as? Int).map(Double.init),
          let h = (canvasDict["height"] as? Double) ?? (canvasDict["height"] as? Int).map(Double.init) else {
      throw ProjectParseError.invalid("invalid canvasSize")
    }
    let fps = (dict["fps"] as? Int) ?? (dict["fps"] as? Double).map { Int($0) } ?? 30
    let durationMs = (dict["durationMs"] as? Double) ?? 0
    let trackArr = (dict["tracks"] as? [[String: Any]]) ?? []
    let tracks = try trackArr.map { try parseTrack($0) }
    return Project(
      id: id, schemaVersion: schemaVersion,
      canvasSize: CGSize(width: w, height: h),
      fps: fps, durationMs: durationMs, tracks: tracks
    )
  }

  private static func parseTrack(_ d: [String: Any]) throws -> ProjectTrack {
    guard let kind = d["kind"] as? String, let id = d["id"] as? String else {
      throw ProjectParseError.invalid("track missing kind/id")
    }
    switch kind {
    case "video":
      let clipDicts = (d["clips"] as? [[String: Any]]) ?? []
      let clips = try clipDicts.map { try parseVideoClip($0) }
      return .video(id: id, clips: clips)
    case "audio":
      let clipDicts = (d["clips"] as? [[String: Any]]) ?? []
      let clips = try clipDicts.map { try parseAudioClip($0) }
      return .audio(id: id, clips: clips)
    case "overlay":
      let itemDicts = (d["items"] as? [[String: Any]]) ?? []
      let items = try itemDicts.map { try parseOverlay($0) }
      return .overlay(id: id, items: items)
    default:
      throw ProjectParseError.invalid("unknown track kind \(kind)")
    }
  }

  private static func parseVideoClip(_ d: [String: Any]) throws -> ProjectVideoClip {
    let id = (d["id"] as? String) ?? UUID().uuidString
    let kind = d["kind"] as? String   // "image" marks an image-clip inside a video track
    let isImage = kind == "image"
    let sourceUri = (d["sourceUri"] as? String) ?? ""
    guard !sourceUri.isEmpty || isImage else { throw ProjectParseError.invalid("video clip \(id) missing sourceUri") }
    let imageDurationMs = d["durationMs"] as? Double
    let sourceRange = parseTimeRange(d["sourceRange"]) ?? ProjectTimeRange(startMs: 0, endMs: imageDurationMs ?? 0)
    guard let timelineRange = parseTimeRange(d["timelineRange"]) else {
      throw ProjectParseError.invalid("video clip \(id) missing timelineRange")
    }
    let transition = parseTransition(d["transition"] as? [String: Any])
    let originalVolume = (d["originalVolume"] as? Double) ?? 1.0
    let resolvedSourceUri = isImage ? (d["sourceUri"] as? String ?? "") : sourceUri
    return ProjectVideoClip(
      id: id, sourceUri: resolvedSourceUri,
      sourceRange: sourceRange, timelineRange: timelineRange,
      transition: transition, originalVolume: originalVolume,
      isImage: isImage, imageDurationMs: imageDurationMs
    )
  }

  private static func parseAudioClip(_ d: [String: Any]) throws -> ProjectAudioClip {
    let id = (d["id"] as? String) ?? UUID().uuidString
    guard let sourceUri = d["sourceUri"] as? String, !sourceUri.isEmpty else {
      throw ProjectParseError.invalid("audio clip \(id) missing sourceUri")
    }
    guard let sourceRange = parseTimeRange(d["sourceRange"]),
          let timelineRange = parseTimeRange(d["timelineRange"]) else {
      throw ProjectParseError.invalid("audio clip \(id) missing ranges")
    }
    let volume = (d["volume"] as? Double) ?? 1.0
    let trimToVideo = (d["trimToVideo"] as? Bool) ?? false
    return ProjectAudioClip(
      id: id, sourceUri: sourceUri,
      sourceRange: sourceRange, timelineRange: timelineRange,
      volume: volume, trimToVideo: trimToVideo
    )
  }

  private static func parseOverlay(_ d: [String: Any]) throws -> ProjectOverlayClip {
    guard let kind = d["kind"] as? String else { throw ProjectParseError.invalid("overlay missing kind") }
    let id = (d["id"] as? String) ?? UUID().uuidString
    let timelineRange = parseTimeRange(d["timelineRange"])
    if kind == "text" {
      return .text(ProjectTextOverlayClip(
        id: id,
        content: (d["content"] as? String) ?? "",
        x: (d["x"] as? Double) ?? 0,
        y: (d["y"] as? Double) ?? 0,
        anchor: (d["anchor"] as? String) ?? "center",
        textAlign: (d["textAlign"] as? String) ?? "center",
        paddingX: (d["paddingX"] as? Double) ?? 12,
        paddingY: (d["paddingY"] as? Double) ?? 8,
        fontSize: (d["fontSize"] as? Double) ?? 32,
        color: (d["color"] as? String) ?? "#FFFFFF",
        fontWeight: (d["fontWeight"] as? String) ?? "normal",
        fontStyle: (d["fontStyle"] as? String) ?? "normal",
        fontFamily: (d["fontFamily"] as? String) ?? "system",
        backgroundColor: d["backgroundColor"] as? String,
        cornerRadius: (d["cornerRadius"] as? Double) ?? 0,
        strokeColor: d["strokeColor"] as? String,
        strokeWidth: (d["strokeWidth"] as? Double) ?? 0,
        shadowColor: d["shadowColor"] as? String,
        shadowRadius: (d["shadowRadius"] as? Double) ?? 0,
        shadowOpacity: (d["shadowOpacity"] as? Double) ?? 1.0,
        highlightColor: d["highlightColor"] as? String,
        highlightStart: (d["highlightStart"] as? Double).map { Int($0) } ?? (d["highlightStart"] as? Int),
        highlightLength: (d["highlightLength"] as? Double).map { Int($0) } ?? (d["highlightLength"] as? Int),
        rotation: (d["rotation"] as? Double) ?? 0,
        timelineRange: timelineRange
      ))
    } else if kind == "image" {
      return .image(ProjectImageOverlayClip(
        id: id,
        uri: (d["uri"] as? String) ?? "",
        x: (d["x"] as? Double) ?? 0,
        y: (d["y"] as? Double) ?? 0,
        width: (d["width"] as? Double) ?? 0.2,
        height: (d["height"] as? Double) ?? 0.2,
        opacity: (d["opacity"] as? Double) ?? 1.0,
        rotation: (d["rotation"] as? Double) ?? 0,
        timelineRange: timelineRange
      ))
    } else {
      throw ProjectParseError.invalid("unknown overlay kind \(kind)")
    }
  }

  private static func parseTimeRange(_ raw: Any?) -> ProjectTimeRange? {
    guard let d = raw as? [String: Any],
          let s = d["startMs"] as? Double,
          let e = d["endMs"] as? Double else { return nil }
    return ProjectTimeRange(startMs: s, endMs: e)
  }

  private static func parseTransition(_ d: [String: Any]?) -> ProjectClipTransition {
    guard let d = d, let type = d["type"] as? String else { return .cut }
    switch type {
    case "fade": return .fade(durationMs: (d["durationMs"] as? Double) ?? 500)
    case "fadeToBlack": return .fadeToBlack(durationMs: (d["durationMs"] as? Double) ?? 500)
    default: return .cut
    }
  }
}
